src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs

```diff
@@ -63,14 +63,100 @@ internal static IEnumerable<RuntimeLibrary> GetCandidateLibraries(DependencyCont
                 return Enumerable.Empty<RuntimeLibrary>();
             }
 
-            return dependencyContext.RuntimeLibraries.Where(IsCandidateLibrary);
+            var candidatesResolver = new CandidatesResolver(dependencyContext.RuntimeLibraries, ReferenceAssemblies);
+            return candidatesResolver.GetCandidates();
         }
 
-        private static bool IsCandidateLibrary(RuntimeLibrary library)
+        private class CandidatesResolver
         {
-            Debug.Assert(ReferenceAssemblies != null);
-            return !ReferenceAssemblies.Contains(library.Name) &&
-                library.Dependencies.Any(dependency => ReferenceAssemblies.Contains(dependency.Name));
+            private readonly IDictionary<string, Dependency> _dependencies;
+
+            public CandidatesResolver(IReadOnlyList<RuntimeLibrary> dependencies, ISet<string> referenceAssemblies)
+            {
+                _dependencies = dependencies
+                    .ToDictionary(d => d.Name, d => CreateDependency(d, referenceAssemblies));
+            }
+
+            private Dependency CreateDependency(RuntimeLibrary library, ISet<string> referenceAssemblies)
+            {
+                var classification = DependencyClassification.Unknown;
+                if (referenceAssemblies.Contains(library.Name))
+                {
+                    classification = DependencyClassification.MvcReference;
+                }
+
+                return new Dependency(library, classification);
+            }
+
+            private DependencyClassification ComputeClasification(string dependency)
+            {
+                Debug.Assert(_dependencies.ContainsKey(dependency));
+
+                var candidateEntry = _dependencies[dependency];
+                if (candidateEntry.Clasification != DependencyClassification.Unknown)
+                {
+                    return candidateEntry.Clasification;
+                }
+                else
+                {
+                    var classification = DependencyClassification.NotCandidate;
+                    foreach (var candidateDependency in candidateEntry.Library.Dependencies)
+                    {
+                        if (classification == DependencyClassification.Candidate)
+                        {
+                            break;
+                        }
+
+                        var dependencyClassification = ComputeClasification(candidateDependency.Name);
+                        if (dependencyClassification == DependencyClassification.Candidate ||
+                            dependencyClassification == DependencyClassification.MvcReference)
```

Comment:
No, that's why there's a specific enum value for it. https://github.com/aspnet/Mvc/pull/4685/files#diff-c1d05e31da7395607e6b98cd37032d1eR96 reference mvc assemblies never change category.

The algorithm works by tagging the mvc reference assembleis as MvcReference and then working its way up the ree marking referencing libraries as candidates.